geomesa-utils/src/main/scala/org/locationtech/geomesa/utils/concurrent/CachedThreadPool.scala

```diff
@@ -0,0 +1,138 @@
+/***********************************************************************
+ * Copyright (c) 2013-2020 Commonwealth Computer Research, Inc.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Apache License, Version 2.0
+ * which accompanies this distribution and is available at
+ * http://www.opensource.org/licenses/apache2.0.php.
+ ***********************************************************************/
+
+package org.locationtech.geomesa.utils.concurrent
+
+import java.util.concurrent._
+import java.util.concurrent.atomic.AtomicBoolean
+import java.util.concurrent.locks.ReentrantLock
+
+import com.google.common.util.concurrent.MoreExecutors
+
+/**
+ * Executor service that will grow up to the number of threads specified.
+ *
+ * Assertion: core java executors offer 3 features that we want, but you can only use two of them at once:
+ *   1. Add threads as needed but re-use idle threads if available
+ *   2. Limit the total number of threads available
+ *   3. Accept and queue tasks once the thread limit is reached
+ *
+ * This executor allows us to do all three. It is backed by an unlimited, cached thread pool that will handle
+ * re-use and expiration of threads, while it tracks the number of threads used. It is fairly lightweight, and
+ * can be instantiated for a given task and then shutdown after use.
+ *
+ * @param maxThreads max threads to use at once
+ */
+class CachedThreadPool(maxThreads: Int) extends AbstractExecutorService {
+
+  @volatile
+  private var available = maxThreads
+  private val queue = new java.util.LinkedList[TrackableFutureTask[_]]()
+  private val tasks = new java.util.HashSet[Future[_]]()
+  private val stopped = new AtomicBoolean(false)
+  private val lock = new ReentrantLock()
+
+  override def shutdown(): Unit = stopped.set(true)
+
+  override def shutdownNow(): java.util.List[Runnable] = {
+    stopped.set(true)
+    lock.lock()
+    try {
+      val waiting = new java.util.ArrayList[Runnable](queue)
+      queue.clear()
+      // copy the running tasks to prevent concurrent modification errors in the synchronous cancel call
+      val running = new java.util.ArrayList[Future[_]](tasks).iterator()
+      while (running.hasNext) {
+        running.next.cancel(true)
+      }
+      waiting
+    } finally {
+      lock.unlock()
+    }
+  }
+
+  override def isShutdown: Boolean = stopped.get
+
+  override def isTerminated: Boolean = stopped.get && available == maxThreads // should be safe to read a volatile primitive
+
+  override def awaitTermination(timeout: Long, unit: TimeUnit): Boolean = {
+    // TODO could probably make a wait instead of a poll
```

Comment:
Should we add a ticket for this?  Or sort it out before mergin?

Reply:
it might be somewhat complicated so i didn't do it at first... i don't think we use this method much/at all, so i'll open a ticket.

Reply:
https://geomesa.atlassian.net/browse/GEOMESA-2795

Reply:
i'll add the ticket number to the comment

Reply:
added